crates/block-producer/src/mempool/batch_graph.rs

```diff
@@ -0,0 +1,344 @@
+        if !duplicates.is_empty() {
+            return Err(BatchInsertError::DuplicateTransactions(duplicates));
+        }
+
+        // Reverse lookup parent batch IDs. Take care to allow for parent transactions within this
+        // batch i.e. internal dependencies.
+        transactions.iter().for_each(|tx| {
+            parents.remove(tx);
+        });
+        let parent_batches = parents
+            .into_iter()
+            .map(|tx| {
+                self.transactions
+                    .get(&tx)
+                    .copied()
+                    .ok_or(BatchInsertError::UnknownParentTransaction(tx))
+            })
+            .collect::<Result<_, _>>()?;
+
+        self.inner.insert_pending(id, parent_batches)?;
+
+        for tx in transactions.iter().copied() {
+            self.transactions.insert(tx, id);
+        }
+        self.batches.insert(id, transactions);
+
+        Ok(())
+    }
+
+    /// Removes the batches and their descendants from the graph.
+    ///
+    /// # Returns
+    ///
+    /// Returns all removes batches and their transactions.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if any of the batches are not currently in the graph.
+    pub fn remove_batches(
+        &mut self,
+        batch_ids: BTreeSet<BatchJobId>,
+    ) -> Result<BTreeMap<BatchJobId, Vec<TransactionId>>, GraphError<BatchJobId>> {
+        // This returns all descendent batches as well.
+        let batch_ids = self.inner.purge_subgraphs(batch_ids)?;
+
+        // SAFETY: These batches must all have been inserted since they are emitted from the inner
+        // dependency graph, and therefore must all be in the batches mapping.
+        let batches = batch_ids
+            .into_iter()
+            .map(|batch_id| {
+                (batch_id, self.batches.remove(&batch_id).expect("batch should be removed"))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        for tx in batches.values().flatten() {
+            self.transactions.remove(tx);
+        }
+
+        Ok(batches)
+    }
+
+    /// Removes the set of committed batches from the graph.
+    ///
+    /// The batches _must_ have been previously selected for inclusion in a block using
+    /// [`select_block`](Self::select_block). This is intended for limiting the size of the graph by
+    /// culling committed data.
+    ///
+    /// # Returns
+    ///
+    /// Returns the transactions of the pruned batches.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if
+    ///   - any batch was not previously selected for inclusion in a block
+    ///   - any batch is unknown
+    ///   - any parent batch would be left dangling in the graph
+    ///
+    /// The last point implies that batches should be removed in block order.
+    pub fn prune_committed(
+        &mut self,
+        batch_ids: BTreeSet<BatchJobId>,
+    ) -> Result<Vec<TransactionId>, GraphError<BatchJobId>> {
+        self.inner.prune_processed(batch_ids.clone())?;
```

Comment:
Sort of. The pruning is atomic, however the overall method loses this property if we move this further down as we'll then have modified internal state.

Though to be honest I'm not sure this property is worth anything to us. In general if we get into this "corrupt" graph state we're probably just screwed.

I'll think about this some more; my earliest iterations had no errors and just panic'd everywhere since there isn't anything we can actually do if the graph is corrupt..

Reply:
I've added a comment at least for this.

Reply:
@bobbinth any thoughts?

Reply:
I'm not too concered about the `clone()` here because the number of batches will always be relatively small (it is capped at 64, but also, it'll be a while till we get to 64). So, I'd probably prefer keeping the method atomic rather than avoid cloning.